File: typescript/src/core/TypeSubscription.ts
import { ISubscriptionDefinition } from "../contracts/ISubscriptionDefinition";
import { AgentId, TopicId } from "../contracts/IAgentRuntime";
import { AgentType } from "../contracts/AgentType";

/**
 * This subscription matches on topics based on the exact type and maps to agents using the source of the topic as the agent key.
 * This subscription causes each source to have its own agent instance.
 * @example
 * ```typescript
 * const subscription = new TypeSubscription("t1", "a1");
 * ```
 * In this case:
 * - A TopicId with type "t1" and source "s1" will be handled by an agent of type "a1" with key "s1"
 * - A TopicId with type "t1" and source "s2" will be handled by an agent of type "a1" with key "s2"
 */
export class TypeSubscription implements ISubscriptionDefinition {
  /**
   * Gets the unique identifier of the subscription.
   */
  public readonly id: string;

  /**
   * The exact topic type used for matching.
   */
  private readonly topicType: string;

  /**
   * The agent type that handles this subscription.
   */
  private readonly agentType: AgentType;

  /**
   * Initializes a new instance of the TypeSubscription class.
   * @param topicType The exact topic type to match against
   * @param agentType Agent type to handle this subscription
   * @param id Optional unique identifier for the subscription. If not provided, a new UUID will be generated
   */
  constructor(topicType: string, agentType: AgentType, id?: string) {
    this.topicType = topicType;
    this.agentType = agentType;
    this.id = id ?? crypto.randomUUID();
  }

  /**
   * Checks if a given TopicId matches the subscription based on an exact type match.
   * @param topic The topic to check
   * @returns true if the topic's type matches exactly, false otherwise
   */
  matches(topic: TopicId): boolean {
    console.log('TypeSubscription.matches:', {
      topicType: topic.type,
      expectedType: this.topicType,
      match: topic.type === this.topicType
    });
    return topic.type === this.topicType;
  }

  /**
   * Maps a TopicId to an AgentId. Should only be called if matches() returns true.
   * @param topic The topic to map
   * @returns An AgentId representing the agent that should handle the topic
   * @throws Error if the topic does not match the subscription
   */
  mapToAgent(topic: TopicId): AgentId {
    if (!this.matches(topic)) {
      throw new Error("TopicId does not match the subscription.");
    }
    console.log('TypeSubscription.mapToAgent:', {
      topic,
      agentType: this.agentType,
      result: { type: this.agentType, key: topic.source } // Use topic.source instead of "default"
    });
    return { 
      type: this.agentType,
      key: topic.source  // Key change: use topic.source instead of hardcoding "default"
    };
  }
}

File: dotnet/src/Microsoft.AutoGen/Core/TypeSubscription.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// TypeSubscription.cs

using System.Diagnostics.CodeAnalysis;

namespace Microsoft.AutoGen.Contracts;

/// <summary>
/// This subscription matches on topics based on the exact type and maps to agents using the source of the topic as the agent key.
/// This subscription causes each source to have its own agent instance.
/// </summary>
/// <remarks>
/// Example:
/// <code>
/// var subscription = new TypeSubscription("t1", "a1");
/// </code>
/// In this case:
/// - A <see cref="TopicId"/> with type `"t1"` and source `"s1"` will be handled by an agent of type `"a1"` with key `"s1"`.
/// - A <see cref="TopicId"/> with type `"t1"` and source `"s2"` will be handled by an agent of type `"a1"` with key `"s2"`.
/// </remarks>
public class TypeSubscription : ISubscriptionDefinition
{
    private readonly string _topicType;
    private readonly AgentType _agentType;
    private readonly string _id;

    /// <summary>
    /// Initializes a new instance of the <see cref="TypeSubscription"/> class.
    /// </summary>
    /// <param name="topicType">The exact topic type to match against.</param>
    /// <param name="agentType">Agent type to handle this subscription.</param>
    /// <param name="id">Unique identifier for the subscription. If not provided, a new UUID will be generated.</param>
    public TypeSubscription(string topicType, AgentType agentType, string? id = null)
    {
        _topicType = topicType;
        _agentType = agentType;
        _id = id ?? Guid.NewGuid().ToString();
    }

    /// <summary>
    /// Gets the unique identifier of the subscription.
    /// </summary>
    public string Id => _id;

    /// <summary>
    /// Gets the exact topic type used for matching.
    /// </summary>
    public string TopicType => _topicType;

    /// <summary>
    /// Gets the agent type that handles this subscription.
    /// </summary>
    public AgentType AgentType => _agentType;

    /// <summary>
    /// Checks if a given <see cref="TopicId"/> matches the subscription based on an exact type match.
    /// </summary>
    /// <param name="topic">The topic to check.</param>
    /// <returns><c>true</c> if the topic's type matches exactly, <c>false</c> otherwise.</returns>
    public bool Matches(TopicId topic)
    {
        return topic.Type == _topicType;
    }

    /// <summary>
    /// Maps a <see cref="TopicId"/> to an <see cref="AgentId"/>. Should only be called if <see cref="Matches"/> returns true.
    /// </summary>
    /// <param name="topic">The topic to map.</param>
    /// <returns>An <see cref="AgentId"/> representing the agent that should handle the topic.</returns>
    /// <exception cref="InvalidOperationException">Thrown if the topic does not match the subscription.</exception>
    public AgentId MapToAgent(TopicId topic)
    {
        if (!Matches(topic))
        {
            throw new InvalidOperationException("TopicId does not match the subscription.");
        }

        return new AgentId(_agentType, topic.Source);
    }

    /// <summary>
    /// Determines whether the specified object is equal to the current subscription.
    /// </summary>
    /// <param name="obj">The object to compare with the current instance.</param>
    /// <returns><c>true</c> if the specified object is equal to this instance; otherwise, <c>false</c>.</returns>
    public override bool Equals([NotNullWhen(true)] object? obj)
    {
        return obj is TypeSubscription other &&
               (Id == other.Id ||
               (AgentType == other.AgentType && TopicType == other.TopicType));
    }

    /// <summary>
    /// Returns a hash code for this instance.
    /// </summary>
    /// <returns>A hash code for this instance, suitable for use in hashing algorithms and data structures.</returns>
    public override int GetHashCode()
    {
        return HashCode.Combine(Id, AgentType, TopicType);
    }
}
